### PR TITLE
 Add Default Targeted Sequencing setting for Projects

### PR DESCRIPTION
**Pull request by Kashif-S**
Jira: GLT-2257

### CODE REVIEW

File: miso-web/src/main/webapp/scripts/hot_dilution.js [hunk 1]
```diff
@@ -76,8 +76,13 @@ HotTarget.dilution = {
               return dil.library.libraryDesignCodeId == code.id;
             }, Constants.libraryDesignCodes);
             if (Utils.array.maybeGetProperty(designCode, 'targetedSequencingRequired')) {
-              setCellMeta('validator', HotUtils.validator.requiredAutocompleteWithNullValue('(None)'));
-              missingValueString = '';
+              setCellMeta('validator', HotUtils.validator.requiredAutocomplete);
+              console.log(dil.library.defaultTargetedSequencing);
```
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/hot_dilution.js]: Please remove the debugging.

**Comment by sqwci** [comment]
SonarQube analysis reported 9 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [Project.java#L47](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/Project.java#L47): Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)
1. ![CRITICAL][CRITICAL] [TargetedSequencing.java#L191](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/TargetedSequencing.java#L191): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultProjectService.java#L165](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L165): Return an empty collection instead of null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [DefaultProjectService.java#L177](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L177): Return an empty collection instead of null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [LimsBindingInitializer.java#L323](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/spring/src/main/java/uk/ac/bbsrc/tgac/miso/spring/LimsBindingInitializer.java#L323): Replace this assert with a proper check. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4274)
1. ![MAJOR][MAJOR] [LimsBindingInitializer.java#L334](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/spring/src/main/java/uk/ac/bbsrc/tgac/miso/spring/LimsBindingInitializer.java#L334): Replace this assert with a proper check. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4274)
1. ![MINOR][MINOR] [ProjectImpl.java#L304](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/ProjectImpl.java#L304): Remove the declaration of thrown exception 'java.lang.SecurityException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultProjectService.java#L265](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L265): Immediately return this expression instead of assigning it to the temporary variable "id". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)
1. ![MINOR][MINOR] [EditProjectController.java#L236](https://github.com/TGAC/miso-lims/blob/13d3a808a1ebfd9fea5b7d1fe02bd6d968feadab/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditProjectController.java#L236): Use isEmpty() to check whether the collection is empty or not. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/TargetedSequencing.java [hunk 2]
```diff
@@ -29,6 +29,14 @@
 @Table(name = "TargetedSequencing")
 public class TargetedSequencing implements Serializable {
 
+  public static final TargetedSequencing NULL = new TargetedSequencing();
+
+  static {
+    NULL.setId(null);
+    NULL.setAlias("No Targeted Sequencing");
```
Comment (callunity) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/TargetedSequencing.java]: Can this be changed to `(None)` like we have in the HoT dropdowns?

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/LibraryDto.java [hunk 3]
```diff
@@ -51,6 +51,7 @@
   private SampleDto sample;
   private String receivedDate;
   private String sampleBoxPositionLabel;
+  private String defaultTargetedSequencing;
```
Comment (callunity) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/LibraryDto.java]: Why is this being added to the library when `targetedSequencing` is a dilution attribute?

File: miso-web/src/main/webapp/scripts/hot_dilution.js [hunk 4]
```diff
@@ -76,8 +76,13 @@ HotTarget.dilution = {
...
+              if(dil.library.defaultTargetedSequencing != null){
+                missingValueString = dil.library.defaultTargetedSequencing;
+              } else {
+                missingValueString = '';
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_dilution.js]: Why `''` as opposed to `'(None)'`?

File: sqlstore/src/main/resources/db/migration/V8801__Default_Targeted_Sequencing.sql [hunk 5]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE Project ADD COLUMN `targetedSequencingId` BIGINT (20) DEFAULT NULL AFTER referenceGenomeId;
+
+ALTER TABLE Project ADD FOREIGN KEY (targetedSequencingId) REFERENCES TargetedSequencing (targetedSequencingId);
```
Comment (callunity) [inline sqlstore/src/main/resources/db/migration/V8801__Default_Targeted_Sequencing.sql]: Syntax is `ALTER TABLE Project ADD CONSTRAINT named_constraint FOREIGN KEY (targetedSequencingId)...`

**Comment by sqwci** [comment]
SonarQube analysis reported 21 issues
* ![CRITICAL][CRITICAL] 11 critical
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 5 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [Project.java#L47](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/Project.java#L47): Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)
1. ![CRITICAL][CRITICAL] [TargetedSequencing.java#L191](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/TargetedSequencing.java#L191): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L551](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L551): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L577](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L577): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L611](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L611): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L811](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L811): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L829](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L829): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L852](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L852): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L914](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L914): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L926](https://github.com/TGAC/miso-lims/blob/f37f416a56de2f7801644051488dc66c389c307e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L926): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/webapp/pages/editProject.jsp [hunk 6]
```diff
@@ -132,6 +132,14 @@
         </form:select>
     </td>
   </tr>
+  <tr>
+    <td>Targeted Sequencing :*</td>
```
Comment (callunity) [inline miso-web/src/main/webapp/pages/editProject.jsp]: `Default Targeted Sequencing`

**Comment by sqwci** [comment]
SonarQube analysis reported 21 issues
* ![CRITICAL][CRITICAL] 11 critical
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 5 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [Project.java#L47](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/Project.java#L47): Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)
1. ![CRITICAL][CRITICAL] [TargetedSequencing.java#L191](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/TargetedSequencing.java#L191): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L551](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L551): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L577](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L577): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L611](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L611): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L811](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L811): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L829](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L829): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L852](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L852): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L914](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L914): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L926](https://github.com/TGAC/miso-lims/blob/d2fe5412010d541ae7307ba0d45c69e6a9eef940/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L926): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**